stdlib/public/Concurrency/AsyncStreamBackpressure.swift

```diff
@@ -0,0 +1,1961 @@
+            let callbackToken = shouldProduceMore ? nil : self.nextCallbackToken()
+
+            self._state = .streaming(.init(
+                backPressureStrategy: initial.backPressureStrategy,
+                iteratorInitialized: initial.iteratorInitialized,
+                onTermination: initial.onTermination,
+                buffer: buffer,
+                consumerContinuation: nil,
+                producerContinuations: .init(),
+                cancelledAsyncProducers: .init(),
+                hasOutstandingDemand: shouldProduceMore
+            ))
+
+            return .init(callbackToken: callbackToken)
+
+        case .streaming(var streaming):
+            self._state = .modify
+
+            streaming.buffer.append(contentsOf: sequence)
+
+            // We have an element and can resume the continuation
+            let shouldProduceMore = streaming.backPressureStrategy.didYield(bufferDepth: streaming.buffer.count)
+            streaming.hasOutstandingDemand = shouldProduceMore
+            let callbackToken = shouldProduceMore ? nil : self.nextCallbackToken()
+
+            if let consumerContinuation = streaming.consumerContinuation {
+                guard let element = streaming.buffer.popFirst() else {
+                    // We got a yield of an empty sequence. We just tolerate this.
+                    self._state = .streaming(streaming)
+
+                    return .init(callbackToken: callbackToken)
+                }
+
+                // We got a consumer continuation and an element. We can resume the consumer now
+                streaming.consumerContinuation = nil
+                self._state = .streaming(streaming)
+                return .init(
+                    callbackToken: callbackToken,
+                    continuationAndElement: (consumerContinuation, element)
+                )
+            } else {
+                // We don't have a suspended consumer so we just buffer the elements
+                self._state = .streaming(streaming)
+                return .init(
+                    callbackToken: callbackToken
+                )
+            }
+
+        case .sourceFinished, .finished:
+            // If the source has finished we are dropping the elements.
+            return .throwFinishedError
+
+        case .modify:
+            fatalError("AsyncStream internal inconsistency")
+        }
+    }
+
+    /// Actions returned by `enqueueProducer()`.
+    enum EnqueueProducerAction {
+        /// Indicates that the producer should be notified to produce more.
+        case resumeProducer((Result<Void, Error>) -> Void)
+        /// Indicates that the producer should be notified about an error.
+        case resumeProducerWithError((Result<Void, Error>) -> Void, Error)
+    }
+
+    @available(SwiftStdlib 9999, *)
+    mutating func enqueueProducer(callbackToken: UInt, onProduceMore: @Sendable @escaping (Result<Void, Error>) -> Void) -> EnqueueProducerAction? {
```

Comment:
would this be better suited with an opaque token? having a bare UInt seems hazardous.

Reply:
This is intentional since it is on the `Storage`/`StateMachine`. We cannot pass a concrete `CallbackToken` type here since we are supporting both the throwing and non-throwing variants with a single `Storage`/`StateMachine` and both have separate types for the `CallbackToken`.